extensions/default/src/validateInstances.ts

```diff
@@ -0,0 +1,319 @@
+  isReconstructable: boolean
+): displayServiceMessageList {
+  const messages = new displayServiceMessageList();
+  if (!instances.length) {
+    messages.addMessage(displaySetMessageCodes.NO_VALID_INSTANCES);
+  }
+
+  const firstInstance = instances[0];
+  if (firstInstance.ImageType.includes('LOCALIZER')) {
+    return messages;
+  }
+
+  const isMultiframe = firstInstance.NumberOfFrames > 1;
+  const Modality = firstInstance.Modality;
+  if (!constructableModalities.includes(Modality)) {
+    return messages;
+  }
+
+  // Can't reconstruct if all instances don't have the ImagePositionPatient.
+  if (
+    !isMultiframe &&
+    !instances.every(instance => instance.ImagePositionPatient)
+  ) {
+    messages.addMessage(displaySetMessageCodes.NO_POSITION_INFORMATION);
+  }
+
+  const sortedInstances = sortInstancesByPosition(instances);
+
+  if (isMultiframe) {
+    checkMultiFrame(sortedInstances[0], messages);
+  } else {
+    checkSingleFrames(sortedInstances, messages);
+  }
+
+  if (!isReconstructable) {
+    messages.addMessage(displaySetMessageCodes.NOT_RECONSTRUCTABLE);
+  }
+  return messages;
+}
+
+/**
+ * Check various multi frame issues. It calls OHIF core functions
+ * @param {*} multiFrameInstance
+ * @param {*} warnings
+ */
+function checkMultiFrame(
+  multiFrameInstance,
+  messages: displayServiceMessageList
+): void {
+  if (!hasPixelMeasurements(multiFrameInstance)) {
+    messages.addMessage(
+      displaySetMessageCodes.MULTIFRAME_NO_PIXEL_MEASUREMENTS
+    );
+  }
+
+  if (!hasOrientation(multiFrameInstance)) {
+    messages.addMessage(displaySetMessageCodes.MULTIFRAME_NO_ORIENTATION);
+  }
+
+  if (!hasPosition(multiFrameInstance)) {
+    messages.addMessage(
+      displaySetMessageCodes.MULTIFRAME_NO_POSITION_INFORMATION
+    );
+  }
+}
+
+/**
+ * Checks if there is a position shift between consecutive frames
+ * @param {*} previousPosition
+ * @param {*} actualPosition
+ * @param {*} scanAxisNormal
+ * @param {*} averageSpacingBetweenFrames
+ * @returns
+ */
+function _checkPositionShift(
+  previousPosition,
+  actualPosition,
+  scanAxisNormal,
+  averageSpacingBetweenFrames
+) {
+  // predicted position should be the previous position added by the multiplication
+  // of the scanAxisNormal and the average spacing between frames
+  const predictedPosition = vec3.scaleAndAdd(
+    vec3.create(),
+    previousPosition,
+    scanAxisNormal,
+    averageSpacingBetweenFrames
+  );
+  return (
+    vec3.distance(actualPosition, predictedPosition) >
+    averageSpacingBetweenFrames
+  );
+}
+
+/**
+ * Check if the frames in a series has different dimensions
+ * @param {*} instances
+ * @returns
+ */
+function checkSeriesDimensions(instances) {
+  const firstImage = instances[0];
+  const firstImageRows = toNumber(firstImage.Rows);
+  const firstImageColumns = toNumber(firstImage.Columns);
+
+  for (let i = 1; i < instances.length; i++) {
+    const instance = instances[i];
+    const { Rows, Columns } = instance;
```

Comment:
I think all these checks should be implemented as separate files and unit tested.